py/malloc: Add m_tracked_realloc().

andrewleech opened the pull request:
### Summary
This PR provides a `m_tracked_realloc()` function to compliment the existing `m_tracked_calloc()` and `m_tracked_free()` functions. 

I'm working on improvements to the LVGL micropython bindings (as user c module) and these functions look safer to provide to that library for dynamic memory support.

I may be wrong here in my assumptions about how the various forms of alloc/free work, but if these are #defined in to be used in a large user c module that expects "standard" malloc/callloc/free function semantics would this tracked version be safer in the sense of ensuring the gc doesn't collect things that might still be in use deep in said C library?

### Testing

The main testing done so far is via the updated coverage test included here. 
I'm intending to test this further in user C modules like LVGL.

### Trade-offs and Alternatives

There's already a non-tracked `m_realloc()` function however when `MICROPY_MALLOC_USES_ALLOCATED_SIZE` is enabled this function requires the previous size to also be provided so no longer matches C standard lib signature. This setting is on by default in the unix port.

andrewleech commented on py/malloc.c:
```diff
@@ -263,6 +263,47 @@ void *m_tracked_calloc(size_t nmemb, size_t size) {
+    #if MICROPY_TRACKED_ALLOC_STORE_SIZE
+    prev_bytes = node->size;
+    #else
+    prev_bytes = gc_nbytes(node);
```
Hi @dpgeorge I copied this logic from `free` below but feel there's a discrepancy here?  When `node->size` is set in alloc above it's set to `nmemb * size` so just the requested allocation size. 
I would have thought the `gc_nbytes(node)` version would return the entire allocated size, ie `sizeof(m_tracked_node_t) + nmemb * size` ?
Is this true, and if so does it not really matter as its only used for metrics reporting? I just found it a little confusing so would be happy to add some clarifying comments if needed, or update the impl to be consistent in both cases.

projectgus replied:
FWIW I looked at this code last week, and my understanding is same for yours - this is only ever used for stats, either `MICROPY_TRACKED_ALLOC_STORE_SIZE` for debugging tracked alloc or complete memory stats when  `MICROPY_MALLOC_USES_ALLOCATED_SIZE` is set.

That said, maybe could we remove one of the config permutations by requiring `MICROPY_TRACKED_ALLOC_STORE_SIZE` to be set if `MICROPY_MALLOC_USES_ALLOCATED_SIZE` is set?

I also toyed with the idea of adding a `inline static size_t get_size(node)` function here to simplify some of these copy-paste ifdefs. I think I decided it was marginal for my PR, but if we're adding one more here it might be worth it!

projectgus commented on py/malloc.c:
```diff
@@ -263,6 +263,49 @@ void *m_tracked_calloc(size_t nmemb, size_t size) {
+        sizeof(m_tracked_node_t) + n_bytes
+        );
+    if (node == NULL) {
+        node->prev->next = node->next;
```
This line dereferences a NULL pointer. You might need to use a second variable for the result, to disambiguate.

(Can also cover these lines in the coverage test by trying to `m_tracked_realloc()` to an unreasonably large size.)

projectgus commented on py/malloc.c:
```diff
@@ -263,6 +263,49 @@ void *m_tracked_calloc(size_t nmemb, size_t size) {
+    node->size = n_bytes;
+    #endif
+    node->prev->next = node;
+    node->next->prev = node;
```
I think `node->next` and `node->prev` can be NULL at the head and tail of the list, respectively.

projectgus reviewed (commented):
This looks potentially useful, for sure! Left a couple of comments inline.

I was worried about the code size impact but I guess this only gets compiled if C code links to `realloc`, yeah? So seems pretty worthwhile given that tradeoff.

projectgus commented on ports/unix/coverage.c:
```diff
@@ -20,6 +20,16 @@
 #include "py/binary.h"
 #include "py/bc.h"
 
+// Forward declaration for tracked allocation testing
```
Re-declaring m_tracked_node_t here seems risky (and it already has potential ABI mismatch as the "real" `size` member is gated on `MICROPY_TRACKED_ALLOC_STORE_SIZE` not `!MICROPY_ENABLE_GC`.

Is it possible to rewrite the tests here so that they test the behaviour that we want to see from `m_tracked_realloc()` as a black box, rather than testing the internals of the implementation?

Perhaps something like: doing some tracked allocations, realloc them, run a GC pass, check everything is still valid, free them, allocate some more?

If it's definitely necessary to test the internals by walking the linked list, maybe we can create a `py/tracked_alloc.h` and put all the relevant declarations there, and then include that header from `py/misc.h` for compatibility?

Big preference for testing this as a black box here, though, if possible.

projectgus reviewed (commented):
Thanks for keeping this ticking along, @andrewleech! How are your experiments with LVGL going?

(I think you're right that it's safer to use this API from LVGL rather than to rely on the GC pointer tracking. An upside is that if we eventually implement [selective collect ala CircuitPython](https://github.com/orgs/micropython/discussions/17142) then all the LVGL buffers would be automatically marked as "no pointers here" and skipped by the GC, which is probably very beneficial for GC times!)

andrewleech commented:
Thanks for the detailed feedback, I've updated the implementation based on your suggestions and rebased onto latest master.

Switched to the alloc+copy+free approach which eliminates the race condition entirely - each operation handles its own locking so there's no window where the list is in an inconsistent state. Also added the NULL check after allocation so if m_tracked_calloc fails the original pointer stays intact per standard realloc semantics.

The tests are now black-box, removed the m_tracked_node_t redeclaration and instead test data preservation during grow/shrink, GC stability after multiple reallocs, and the edge cases like realloc(NULL, n) and realloc(ptr, 0). The FLIP_POINTER test now always shows location_changed=1 since alloc+copy+free always moves the pointer, which is expected behaviour for this implementation.

Removed the unrelated run-tests.py documentation changes as you suggested.

The alloc+copy+free approach trades some efficiency for simplicity - always allocates a new buffer rather than potentially growing in-place, which increases peak memory pressure briefly while both buffers exist. The correctness and simplicity gains are worth it though.

The get_tracked_node_size() call happens outside the lock but that's safe since node->size is write-once and gc_nbytes reads stable metadata for live allocations, same pattern as m_tracked_free uses.

Known gap is the allocation failure path isn't tested since that would need deterministic heap exhaustion which is tricky in the coverage tests.

Commit messages are probably a bit verbose for micropython conventions, happy to trim if you think it's worth it. Also aware the m_tracked_count_links byte counting has a pre-existing inconsistency between the STORE_SIZE and gc_nbytes paths but that's unrelated to this PR.

projectgus commented on py/malloc.c:
```diff
@@ -292,18 +301,47 @@ void *m_tracked_calloc(size_t nmemb, size_t size) {
+    }
+
+    // Copy data (minimum of old and new size)
+    size_t copy_size = (old_size < n_bytes) ? old_size : n_bytes;
```
Nitpick, but can use macro here:

```suggestion
    size_t copy_size = MIN(old_size, n_bytes);
```

projectgus reviewed (approved):
Thanks @andrewleech. It's certainly much simpler code now! I hope I didn't bulldoze you into this approach, and it still meets your needs for this API.

pi-anl committed:
py/malloc: Add m_tracked_realloc.

Add realloc support to the tracked allocation API, using an alloc-copy-free
approach for simplicity and thread safety.  This function compliments the
existing `m_tracked_calloc()` and `m_tracked_free()` functions.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

pi-anl committed:
tests/ports/unix: Add tests for m_tracked_realloc.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>
